File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/validators/JavaContainsJmlConstruct.java
package com.github.jmlparser.validators;

import com.github.javaparser.ast.Jmlish;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.jml.body.JmlBodyDeclaration;
import com.github.javaparser.ast.jml.stmt.JmlStatement;
import com.github.javaparser.ast.jml.stmt.JmlStatements;
import com.github.javaparser.ast.validator.ProblemReporter;
import com.github.javaparser.ast.validator.Validator;

import java.util.function.Predicate;

/**
 * @author Alexander Weigl
 * @version 1 (19.02.22)
 */
public class JavaContainsJmlConstruct implements Validator {
    @Override
    public void accept(Node node, ProblemReporter problemReporter) {
        accept(node, false, problemReporter);
    }

    private void accept(Node current, Boolean inJml, ProblemReporter problemReporter) {
        Predicate<Node> openJml = (Node it) ->
                it instanceof JmlBodyDeclaration ||
                        it instanceof JmlStatement ||
                        it instanceof JmlStatements;

        if (!inJml && (current instanceof Jmlish) && !openJml.test(current)) {
            problemReporter.report(current, "Jml construct used in Java part");
            return;
        }

        for (Node it : current.getChildNodes()) {
            accept(it, inJml || openJml.test(current), problemReporter);
        }
    }
}

File: tools/lint/src/main/kotlin/io/github/jmltoolkit/lint/rules/JavaContainsJmlConstruct.kt
package io.github.jmltoolkit.lint.rules

import com.github.javaparser.ast.Jmlish
import com.github.javaparser.ast.Node
import com.github.javaparser.ast.jml.clauses.JmlContract
import com.github.javaparser.ast.jml.stmt.JmlStatement
import com.github.javaparser.ast.validator.ProblemReporter
import java.util.function.Predicate

/**
 * @author Alexander Weigl
 * @version 1 (19.02.22)
 */
class JavaContainsJmlConstruct : com.github.javaparser.ast.validator.Validator {
    override fun accept(node: Node, problemReporter: ProblemReporter) {
        accept(node, false, problemReporter)
    }

    private fun accept(current: Node, inJml: Boolean, problemReporter: ProblemReporter) {
        val openJml: Predicate<Node> = Predicate<Node> { it: Node? -> it is JmlStatement || it is JmlContract}

        if (!inJml && (current is Jmlish) && !openJml.test(current)) {
            problemReporter.report(current, "Jml construct used in Java part")
            return
        }

        for (it in current.childNodes) {
            accept(it, inJml || openJml.test(current), problemReporter)
        }
    }
}
